### ISSUE TITLE
Extensive debug records of 10035 in recv() 

### ISSUE THREAD

**Issue by TheSil** [giampaolo/pyftpdlib#455]
On Windows when there are no data to be receieved, the WSAEWOULDBLOCK (10035) error is returned from recv() call. The pyftpdlib treats this is correctly as a retry, but unfortunately logs this as an error, because of this:

```
            debug("call: recv(), err: %s" % err, inst=self)
            if err.errno in _ERRNOS_DISCONNECTED:
                self.handle_close()
                return b''
            elif err.errno in _ERRNOS_RETRY:
                raise RetryError
            else:
                raise
```

I believe the debug message should be filtered for the would block case on windows, since the logs are otherwised quite spammed, and because it is a correct case. Locally I've fixed this by this:

```
if os.name != 'nt' or err.errno != 10035:
    debug("call: recv(), err: %s" % err, inst=self)
```

It seems to work, but perhaps not the proper fix. I am reproducing this quite often by just running multiple FTP Servers in multiple threads.

Some references:
https://stackoverflow.com/questions/17064069/recvfrom-error-10035-using-non-blocking-sockets/17064225

**Comment by giampaolo** [comment]
Mmmm does this occur very quickly different times in a row? Because the real problem here is `select()`: it signals the fd as "readable" but it's not (which is expected). What is not expected is that it is not readable for a long time.

**Comment by TheSil** [comment]
It occurs once before commands, but It seems to be really somehow related to have running multiple FTP Server on multiple ip addresses, which I guess is quite specific... Attached an example which fails for me, together with the failed output.
[test.zip](https://github.com/giampaolo/pyftpdlib/files/1738424/test.zip)

I am basically having installed the Microsoft loopback driver and added the 192.168.2.2 and 192.168.200.2 addresses to it. I guess the interesting part is:

```
DEBUG:pyftpdlib:192.168.2.2:59784-[] <- USER anonymous
DEBUG:pyftpdlib:[debug] call: recv(), err: [WinError 10035] Neblokující operaci knihovny soketů nelze dokončit ihned (<FTPHandlerMulti(id=31822704, addr='192.168.2.2:59784')>)
DEBUG:pyftpdlib:192.168.2.2:59784-[] -> 331 Username ok, send password.
DEBUG:pyftpdlib:192.168.2.2:59784-[anonymous] <- PASS ******
DEBUG:pyftpdlib:[debug] call: recv(), err: [WinError 10035] Neblokující operaci knihovny soketů nelze dokončit ihned (<FTPHandlerMulti(id=31822704, addr='192.168.2.2:59784', user='anonymous')>)
DEBUG:pyftpdlib:192.168.2.2:59784-[anonymous] -> 230 Login successful.
INFO:pyftpdlib:192.168.2.2:59784-[anonymous] USER 'anonymous' logged in.
DEBUG:pyftpdlib:192.168.2.2:59784-[anonymous] <- PORT 192,168,2,2,233,138
DEBUG:pyftpdlib:[debug] call: recv(), err: [WinError 10035] Neblokující operaci knihovny soketů nelze dokončit ihned (<FTPHandlerMulti(id=31822704, addr='192.168.2.2:59784', user='anonymous')>)
DEBUG:pyftpdlib:192.168.2.2:59784-[anonymous] -> 200 Active data connection established.
DEBUG:pyftpdlib:192.168.2.2:59784-[anonymous] <- NLST
DEBUG:pyftpdlib:[debug] call: recv(), err: [WinError 10035] Neblokující operaci knihovny soketů nelze dokončit ihned (<FTPHandlerMulti(id=31822704, addr='192.168.2.2:59784', user='anonymous')>)
```

**Comment by TheSil** [comment]
I have checked it again, this time by looking into Cython socket library itself, here where the error is (not) handled:

```
if (s->sock_timeout > 0
            && (CHECK_ERRNO(EWOULDBLOCK) || CHECK_ERRNO(EAGAIN))) {
            /* False positive: sock_func() failed with EWOULDBLOCK or EAGAIN.

               For example, select() could indicate a socket is ready for
               reading, but the data then discarded by the OS because of a
               wrong checksum.

               Loop on select() to recheck for socket readyness. */
            continue;
        }

        /* sock_func() failed */
        if (!err)
            s->errorhandler();
        /* else: err was already set before */
        return -1;
```
Since sock_timeout is 0, it gets to -1 and we see the EWOULDBLOCK. But it seems to be intentional. What is strange is that select was called shortly before that, yet we get EWOULDBLOCK. I have done some more searching and found https://stackoverflow.com/questions/911565/select-recv-and-ewouldblock-on-non-blocking-sockets/911881 . They mention this behavior not that uncommon on some systems and especially on loopback drivers, which is exactly my case. 

Would it make sense to surpress the error message in this combination (nonblocking socket, select() returns data, recv() raises EWOULDBLOCK)? Another option would be to let Cython handle this case, although I don't think they have the knowledge that select was called shortly before that, so they probably cannot distinguish this case...

**Comment by TheSil** [comment]
Also to your original question, it actually occurs exactly once always, and always shortly after select(). I have observed that slight changes in timing (attaching debugger, adding breakpoints) might or might not get rid of that issue totally.

**Comment by TheSil** [comment]
Okay found the problem, it was on my side. Basically I did the 
```
def create_ftp_handler():
    class FTPHandlerMulti(FTPHandler):
      pass
    return FTPHandlerMulti
```
to fight the fact that FTPHandler class has some globals, but I did not want them to be shared amongst the threads. This worked, but there is also the ioloop that was shared in this case. So in fact there was same ioloop object running from two threads, caused this issues (it's a wonder it actually worked somehow...). I have replace the

`self.server = FTPServer(address, handler)`

with

`self.server = FTPServer(address, handler, ioloop=Select())`

to ensure those threads will each have their own loop, and now it works. So we can close this issue.
